conversions for Eigen::Isometry3d in ROS 2

vmayoral opened the pull request:
This commits fetch https://github.com/ros/geometry2/pull/335 into ROS 2 and addresses the requests from https://github.com/ros-planning/moveit2/issues/32 originated at https://github.com/ros-planning/moveit2/pull/12

rhaschke committed:
improve computation efficiency

rhaschke committed:
improve computation efficiency

rhaschke committed:
improve comments

rhaschke committed:
improve comments

rhaschke committed:
normalize quaternions to be in half-space w >= 0

... as in tf1

rhaschke committed:
normalize quaternions to be in half-space w >= 0

... as in tf1

rhaschke committed:
add Eigen::Isometry3d conversions

Fix conflicts and adapt code to ROS 2.0
Edited by @vmayoral.

rhaschke committed:
add Eigen::Isometry3d conversions

Fix conflicts and adapt code to ROS 2.0
Edited by @vmayoral.

vmayoral committed:
Port tf2_kdl

vmayoral committed:
Adapt for eigen3

vmayoral committed:
Return to eigen in package.xml

Follow advice of @tfoote

vmayoral committed:
Remove cmake_modules

vmayoral committed:
Include the right headers

vmayoral committed:
Include the right headers

vmayoral committed:
tf2_eigen, leftover from the cherry-pick

While cherry-picking changes to get isometry3d in

vmayoral committed:
tf2_eigen, leftover from the cherry-pick

While cherry-picking changes to get isometry3d in

vmayoral committed:
Fix problem with stamp

Picked also a leftover from the ROS2 conversion

vmayoral committed:
Fix problem with stamp

Picked also a leftover from the ROS2 conversion

vmayoral committed:
Fix more issues while porting to ROS2

vmayoral committed:
Fix more issues while porting to ROS2

vmayoral committed:
Remove duplicated definitions

vmayoral committed:
Remove duplicated definitions

vmayoral committed:
Undo changes in unit tests

Leave code as it was left by @clalancette.
Only added some additional lines to test also
Isometry as per the commits of Robert

vmayoral committed:
Undo changes in unit tests

Leave code as it was left by @clalancette.
Only added some additional lines to test also
Isometry as per the commits of Robert

davetcoleman commented:
@rhaschke please review since i believe this was originally yours

rhaschke commented on tf2_eigen/include/tf2_eigen/tf2_eigen.h:
```diff
@@ -34,19 +34,33 @@
 #include <Eigen/Geometry>
 #include <geometry_msgs/msg/point_stamped.hpp>
 #include <geometry_msgs/msg/pose_stamped.hpp>
+#include <geometry_msgs/msg/quaternion_stamped.hpp>
```
Aren't these pulled in by the previous headers?

tfoote replied:
Yes, but if we use them in this file we should include them here too.

rhaschke commented on tf2_eigen/include/tf2_eigen/tf2_eigen.h:
```diff
@@ -174,6 +209,99 @@ void doTransform(const Eigen::Affine3d& t_in,
+ */
+inline
+geometry_msgs::msg::Quaternion toMsg(const Eigen::Quaterniond& in) {
+ geometry_msgs::msg::Quaternion msg;
```
There is only one space of indent. I think ROS style requires two.

tfoote replied:
+1

vmayoral replied:
390955226773e4de5a6143338e632425071d896c

rhaschke commented on tf2_eigen/include/tf2_eigen/tf2_eigen.h:
```diff
@@ -186,10 +314,42 @@ geometry_msgs::msg::Pose toMsg(const Eigen::Affine3d& in) {
-  msg.orientation.y = Eigen::Quaterniond(in.rotation()).y();
-  msg.orientation.z = Eigen::Quaterniond(in.rotation()).z();
-  msg.orientation.w = Eigen::Quaterniond(in.rotation()).w();
+  Eigen::Quaterniond q(in.linear());
```
If that's possible, flip the quaternion in a single line operation, instead of manually doing it below.
```suggestion
  Eigen::Quaterniond q(in.linear());
  if (q.w() < 0)
    q *= 1.0;
```

tfoote replied:
I'm not sure that we want to be doing another operation if we can avoid it. Every extra operation is potentially a loss of precision as well as costing resources. 

The negative w values are completely valid quaternions. I'd rather not just normalize it here. This breaks the expectation that fromMsg(toMsg(...)) will return the same value.

rhaschke reviewed (commented):
I'm not in ROS2, so I cannot judge the corresponding changes. The Eigen stuff looks reasonable.

Commit:
Use node inteface pointers in the MessageFilter class

In the MessageFilter class, use node interfaces instead of using the node directly
so that the code works with either rclcpp::Node or rclcpp_lifecycle::LifecycleNode.
Retain the existing node-based interface for backwards compatibility.

Commit:
Update comments

davetcoleman commented:
closes https://github.com/ros-planning/moveit2/issues/32

ivanpauno committed:
Using ament_target_dependencies where possible (#98)

* Fixed ros2#658

Signed-off-by: ivanpauno <ivanpauno@ekumenlabs.com>

* Modified Eigen include directory again

Signed-off-by: ivanpauno <ivanpauno@ekumenlabs.com>

* Deleted custom variable 'Eigen_INCLUDE_DIR'

Signed-off-by: ivanpauno <ivanpauno@ekumenlabs.com>

* revert unrelated whitespace change

Signed-off-by: ivanpauno <ivanpauno@ekumenlabs.com>

* Added Eigen message again

Signed-off-by: ivanpauno <ivanpauno@ekumenlabs.com>

nuclearsandwich committed:
0.11.0

Karsten1987 committed:
Use node interfaces for static transform broadcaster (#104)

* use node interfaces

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

* do not store a copy of topic interface

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

* initialization style

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

vmayoral commented:
@tfoote, what do we need to do to get this merged? it'd be great to get this in for Dashing.

tfoote commented on tf2_eigen/include/tf2_eigen/tf2_eigen.h:
```diff
@@ -186,10 +314,42 @@ geometry_msgs::msg::Pose toMsg(const Eigen::Affine3d& in) {
+  msg.orientation.y = q.y();
+  msg.orientation.z = q.z();
+  msg.orientation.w = q.w();
+  if (msg.orientation.w < 0) {
```
I'd rather not normalize the direction here either.

vmayoral replied:
906520c067b7feffba36c3467b05295df653cbb3

vmayoral committed:
ROS style, double whitespace indentation

vmayoral committed:
ROS style, double whitespace indentation

vmayoral committed:
Remove normalization of the direction

Follows from https://github.com/ros2/geometry2/pull/93/files/66ecb3a34c358b3f9296d25db47672f968e43c46#r276819825
Rationale provided at https://github.com/ros2/geometry2/pull/93/files/66ecb3a34c358b3f9296d25db47672f968e43c46#r276819658

vmayoral committed:
Remove normalization of the direction

Follows from https://github.com/ros2/geometry2/pull/93/files/66ecb3a34c358b3f9296d25db47672f968e43c46#r276819825
Rationale provided at https://github.com/ros2/geometry2/pull/93/files/66ecb3a34c358b3f9296d25db47672f968e43c46#r276819658

vmayoral commented:
@tfoote changes requested apply, would it be ok to merge it?

scheunemann committed:
also export Eigen3 include directories (#102)

tfoote reviewed (commented):
This is really hard to review with the changes basically intermingled with #90 and it looks like some of the fixes for that have been merged here but not there. 

If we could isolate the commits here for review that would be helpful as well as working to rebase/fix the changes in #90 so we don't see them here that would be great. For the short term, this could even target the branch for #90 maybe.

vmayoral commented:
@tfoote changes in tf2_kdl are related to this commit https://github.com/ros2/geometry2/pull/93/commits/b75a4f4a8266cc832f1607b4530b0b69d03cf749  .

AFAIK, removing this commit form this PR should do. Can you confirm if this is what you're requesting above and if there's anything else required?

This way, tf2_kdl changes will be pushed to #90 which @ahcorde is currently reviewing from our side.

emersonknapp committed:
Fix call to create_publisher after API changed (#105)

* Fix call to create_publisher after API changed

Signed-off-by: Emerson Knapp <eknapp@amazon.com>

* revert whitespace changes

mm318 committed:
add LICENSE file (#106)

* add LICENSE file

Signed-off-by: Miaofei <miaofei@amazon.com>

wjwwood committed:
changes to avoid deprecated API's (#107)

Signed-off-by: William Woodall <william@osrfoundation.org>

Karsten1987 committed:
use node interfaces throughout tf2_ros (#108)

* Use node interface pointers in the TransformListener class

In the TransformListener class, use node interfaces instead of using the node directly
so that the code works with either rclcpp::Node or rclcpp_lifecycle::LifecycleNode.
Retain the existing node-based interface for backwards compatibility.

* Format code based on ament_cpplint and ament_uncrustify

* use node interfaces consistently

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

* adher to new rclcpp API

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

* remove std::forward

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

* export symbols for windows

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

* callback also has to be public

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

* address review

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

* refactor node wrapper class

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

* restore latched todo

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

Commit:
Tf2_msgs including actions (#109)

* Adding tf2_msgs actions

tfoote committed:
updating changelog

tfoote committed:
0.11.1

vmayoral committed:
Port tf2_kdl (#90)

* improve computation efficiency

* improve comments

* normalize quaternions to be in half-space w >= 0

... as in tf1

* add Eigen::Isometry3d conversions

Fix conflicts and adapt code to ROS 2.0
Edited by @vmayoral.

* Port tf2_kdl

* Adapt for eigen3

* Return to eigen in package.xml

Follow advice of @tfoote

* Remove cmake_modules

* Update tf2_kdl messages

* Include the right headers

* tf2_eigen, leftover from the cherry-pick

While cherry-picking changes to get isometry3d in

* Fix problem with stamp

Picked also a leftover from the ROS2 conversion

* Fix more issues while porting to ROS2

* Remove duplicated definitions

* Undo changes in unit tests

Leave code as it was left by @clalancette.
Only added some additional lines to test also
Isometry as per the commits of Robert

* Update tf2_eigen, add toMsg2

Convert a Eigen::Vector3d type to a geometry_msgs::msg::Vector3
while avoiding overloading issues with previous definitions

* make tf2_kdl an ament pkg

* Default to C++14

* First steps towards getting tests

* Fix timePointFromTime conversions

* Make unit test compile

* Fixing test tf2_kdl

* including it own includes in tf2_kdl

* Using timeFromTimePoint function from tf2_ros2

* Fixing tf2_kdl test memory leaks

* indent return

* Define _USE_MATH_DEFINES so Windows gets M_PI symbol.

* Quaternion one to one mapping

* Removing delete of a unique_ptr object

* Fixing possible loss of data in windows

* Remove unused NSEC_PER_SECOND.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

* Adding ament_python_install_package to tf2_kdl

* Remove deprecated tf2_kdl/include/tf2_kdl/tf2_kdl subdirectory

* Porting tf2_kdl python test

* Porting tf2_kdl pythonn library

* Cleanup the package.xml and CMakeLists.txt

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

vmayoral commented:
#90 has been merged but this PR still remains. @ahcorde, how do you want to proceed about this? AFAIK, these changes are needed to compile http://github.com/acutronicrobotics/moveit2

clalancette commented:
@vmayoral The thing to do here is to rebase this on top of the current `ros2` branch.  Then this will be much easier to review.  Thanks!

vmayoral committed:
Merge branch 'isometry3d' of https://github.com/AcutronicRobotics/geometry2 into isometry3d

vmayoral commented:
> @vmayoral The thing to do here is to rebase this on top of the current ros2 branch. Then this will be much easier to review. Thanks!

Thanks @clalancette, I was expecting @ahcorde to take care of it since he's ensuring everything's aligned for the Dashing release but I went ahead and did it myself. @ahcorde can you have a look yourself at the changes as well please?

ahcorde committed:
tf2_eigen toMsg return the same value

tfoote commented:
This rebase looks like it has ended up with a lot of duplicate already merged content in the diff. It's over 1000 lines of diff and includes 44 commits several of which are in the last release or two including things like the changelogs making this diff indecipherable.  Which are the commits in this branch that are adding new content to be reviewed?

I pulled it and computed the diff to look like this:

```
$ git diff ros2
diff --git a/tf2_eigen/include/tf2_eigen/tf2_eigen.h b/tf2_eigen/include/tf2_eigen/tf2_eigen.h
index 8f8fd2e..9213915 100644
--- a/tf2_eigen/include/tf2_eigen/tf2_eigen.h
+++ b/tf2_eigen/include/tf2_eigen/tf2_eigen.h
@@ -50,7 +50,7 @@ namespace tf2
  */
  inline
  Eigen::Isometry3d transformToEigen(const geometry_msgs::msg::Transform& t) {
-   return Eigen::Isometry3d(Eigen::Translation3d(t.translation.x, t.translation.y, t.translation.z)
+ return Eigen::Isometry3d(Eigen::Translation3d(t.translation.x, t.translation.y, t.translation.z)
                         * Eigen::Quaterniond(t.rotation.w, t.rotation.x, t.rotation.y, t.rotation.z));
 }
 
@@ -224,12 +224,12 @@ void doTransform(const Eigen::Isometry3d& t_in,
  */
 inline
 geometry_msgs::msg::Quaternion toMsg(const Eigen::Quaterniond& in) {
- geometry_msgs::msg::Quaternion msg;
- msg.w = in.w();
- msg.x = in.x();
- msg.y = in.y();
- msg.z = in.z();
- return msg;
+  geometry_msgs::msg::Quaternion msg;
+  msg.w = in.w();
+  msg.x = in.x();
+  msg.y = in.y();
+  msg.z = in.z();
+  return msg;
 }
 
 /** \brief Convert a Quaternion message type to a Eigen-specific Quaterniond type.
@@ -341,22 +341,6 @@ geometry_msgs::msg::Pose toMsg(const Eigen::Isometry3d& in) {
   return msg;
 }
 
-/** \brief Convert a Eigen::Vector3d type to a geometry_msgs::msg::Vector3.
- * This function is a specialization of the toMsg template defined in tf2/convert.h.
- * The function was renamed to toMsg2 to avoid overloading conflicts with definitions above.
- *
- * \param in The Eigen::Vector3d to convert.
- * \return The Eigen::Vector3d converted to a geometry_msgs::msg::Vector3 message.
- */
-inline
-geometry_msgs::msg::Vector3 toMsg2(const Eigen::Vector3d& in) {
-  geometry_msgs::msg::Vector3 msg;
-  msg.x = in(0);
-  msg.y = in(1);
-  msg.z = in(2);
-  return msg;
-}
-
 /** \brief Convert a Pose message transform type to a Eigen Affine3d.
  * This function is a specialization of the toMsg template defined in tf2/convert.h.
  * \param msg The Pose message to convert.
```

dirk-thomas commented:
@vmayoral Friendly ping.

vmayoral commented:
That's right @tfoote, sorry. Thanks for the ping @dirk-thomas.

Following at https://github.com/ros2/geometry2/pull/113.
